### PR TITLE
LPS-182876 Add menu option for copy in Folders

### PR DESCRIPTION
**Pull request by marcogalluzzi**
According [LPS-182876](https://issues.liferay.com/browse/LPS-182876) I've added a new action in D&M context menu for Folders:

* Under `feature.flag.LPS-182512`
* It is shown as "Copy To"
* It navigates to a new page where the destination folder can be selected
* The destination folder can be located in a different repository
* If the destination folder is in the same repository, it doesn't allow copying a folder into itself.

I'm currently working on adding unit tests for new Copy method of DLAppService.

### CODE REVIEW

**Comment by liferay-lima** [comment]
ci:test:sf

**Comment by liferay-lima** [comment]
ci:test:relevant

File: modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/display/context/DLCopyFolderDisplayContext.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+		portletDisplay.setURLBack(getRedirect());
+
+		if (_liferayPortletResponse instanceof RenderResponse) {
+			RenderResponse renderResponse =
+				(RenderResponse)_liferayPortletResponse;
+
+			renderResponse.setTitle(
+				LanguageUtil.get(_httpServletRequest, "copy-to"));
+		}
+	}
```
Comment (adolfopa) [inline modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/display/context/DLCopyFolderDisplayContext.java:172]: I don't think mutating the request in the constructor is a good idea. I'd either leave the code in the JSP (it is ugly, but it is the traditional way), or expose this method and call it explicitly from the JSP.

**Review by adolfopa — CHANGES_REQUESTED** [review]

File: modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/display/context/DLCopyFolderDisplayContext.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+		return _sourceFolderId;
+	}
+
+	public String getSourceFolderName() {
+		if (_sourceFolderName != null) {
+			return _sourceFolderName;
+		}
+
+		_sourceFolderName = "Home";
```
[the remark this answers is not in the document]
Comment (marcogalluzzi) [reply modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/display/context/DLCopyFolderDisplayContext.java]: Done in commit [LPS-182876 minor improvements](https://github.com/liferay-lima/liferay-portal/pull/4281/commits/2536104f0d5201b0990dea10509f0725a37a8f21)

File: modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/display/context/DLCopyFolderDisplayContext.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+		long groupId = repositoryId;
+
+		Repository repository = RepositoryLocalServiceUtil.fetchRepository(
+			repositoryId);
+
+		if (repository != null) {
+			groupId = repository.getGroupId();
+		}
+
+		return GroupLocalServiceUtil.getGroup(groupId);
```
[the remark this answers is not in the document]
Comment (marcogalluzzi) [reply modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/display/context/DLCopyFolderDisplayContext.java]: Done in commit [LPS-182876 minor improvements](https://github.com/liferay-lima/liferay-portal/pull/4281/commits/2536104f0d5201b0990dea10509f0725a37a8f21)

File: modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/display/context/DLCopyFolderDisplayContext.java (hunk 1 above)
Comment (marcogalluzzi) [reply modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/display/context/DLCopyFolderDisplayContext.java]: I also didn't like the approach I took. I copied it from another place, I supposed it was a "standard" 😓 Now I've exposed the method `setViewAttributes`. I've also changed the required parameters for the `DLCopyFolderDisplayContext` constructor.

See the commit [LPS-182876 minor improvements](https://github.com/liferay-lima/liferay-portal/pull/4281/commits/2536104f0d5201b0990dea10509f0725a37a8f21)

**Comment by marcogalluzzi** [comment]
Hi @adolfopa I've made the requested changes and I've also added tests for the new folder Copy method. Please review last 2 commits only

**Comment by marcogalluzzi** [comment]
ci:test:sf

**Comment by marcogalluzzi** [comment]
ci:test:relevant

**Review by adolfopa — APPROVED** [review]

**Comment by adolfopa** [comment]
ci:forward

**Comment by WYZ-Issac** [comment]
Hi @marcogalluzzi, the above failures are nor related to your PR. I will forward this directly.

**Comment by WYZ-Issac** [comment]
ci:forward:force

**Pull request closed without merging**